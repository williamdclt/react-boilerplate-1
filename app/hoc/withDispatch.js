import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

const withDispatch = (actionCreator, ...args) => WrappedComponent => {
  class Wrapper extends React.Component {
    componentDidMount() {
      this.props.actionDispatcher(...args);
    }

    render() {
      const { actionDispatcher, ...otherProps } = this.props;
      return <WrappedComponent {...otherProps} />;
    }
  }

  Wrapper.propTypes = {
    actionDispatcher: PropTypes.func,
  };

  const mapDispatchToProps = {
    actionDispatcher: actionCreator,
  };

  return connect(
    null,
    mapDispatchToProps,
  )(Wrapper);
};

export default withDispatch;
