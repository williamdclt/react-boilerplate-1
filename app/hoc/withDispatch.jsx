import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

const withDispatch = (actionCreator, ...args) => WrappedComponent => {
  class Wrapper extends React.Component {
    componentDidMount() {
      this.props.hoc_actionDispatcher(...args);
    }

    render() {
      const { hoc_actionDispatcher, ...otherProps } = this.props;
      return <WrappedComponent {...otherProps} />;
    }
  }

  Wrapper.propTypes = {
    hoc_actionDispatcher: PropTypes.func,
  };

  const mapDispatchToProps = {
    hoc_actionDispatcher: actionCreator,
  };

  return connect(null, mapDispatchToProps)(Wrapper);
};

export default withDispatch;
