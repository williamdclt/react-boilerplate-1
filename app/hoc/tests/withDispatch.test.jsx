import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import withDispatch from '../withDispatch';

describe('withDispatch', () => {
  it('dispatches an action when the component mounts', () => {
    const actionCreator = jest.fn();
    actionCreator.mockReturnValue({ type: 'mocked_action' });

    const Wrapped = props => <span {...props} />;

    mount(
      <Provider store={createStore(() => {})}>
        {React.createElement(withDispatch(actionCreator)(Wrapped))}
      </Provider>,
    );
    expect(actionCreator.mock.calls.length).toBe(1);
  });

  it('render the wrapped component', () => {
    const actionCreator = jest.fn();
    actionCreator.mockReturnValue({ type: 'mocked_action' });

    const Wrapped = props => <span {...props} />;

    const component = mount(
      <Provider store={createStore(() => {})}>
        {React.createElement(
          withDispatch(actionCreator)(() => <Wrapped key="overriden_prop" />),
        )}
      </Provider>,
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
