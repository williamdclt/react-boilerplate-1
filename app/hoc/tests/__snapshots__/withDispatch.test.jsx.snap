// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withDispatch render the wrapped component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(Symbol.observable): [Function],
    }
  }
>
  <Connect(Wrapper)>
    <Wrapper
      actionDispatcher={[Function]}
    >
      <Component>
        <Wrapped
          key="overriden_prop"
        >
          <span />
        </Wrapped>
      </Component>
    </Wrapper>
  </Connect(Wrapper)>
</Provider>
`;
